main.go

```diff
@@ -133,6 +137,46 @@ func (d *Downloader) downloadFileWithTimeout(userCtx context.Context, u string)
 	}
 }
 
+func (d *Downloader) getSize(ctx context.Context, u string) (uint64, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodHead, u, nil)
+	if err != nil {
+		return 0, fmt.Errorf("error creating the request for %s: %w", u, err)
+	}
+	resp, err := d.client.Do(req)
+	if err != nil {
+		return 0, fmt.Errorf("error sending a get http request to %s: %w", u, err)
+	}
+	if resp.StatusCode != 200 {
+		return 0, fmt.Errorf("status code is not valid: %v", resp.StatusCode)
+	}
+	defer resp.Body.Close()
+	if resp.ContentLength != 0 && resp.ContentLength != -1 {
+		return uint64(resp.ContentLength), nil
+	} else if resp.Header.Get("Content-Range") != "" {
+		var size uint64
+		contentRangeSplit := strings.Split(resp.Header.Get("Content-Range"), "/")
+		fmt.Sscan(contentRangeSplit[len(contentRangeSplit) -1], &size)
+		return uint64(size), nil
+	}
+	return 0, nil
+}
+
+func (d *Downloader) emitContentSize(ctx context.Context, url, path string, ch chan DownloadStatus) (err error) {                       
+	s := DownloadStatus{
+		URL: url,
+		DownloadedFilePath: path,
+		DowloadedFileSize: 0,
+	}
+	defer func() { ch <- s }()
+	ds, err := d.getSize(ctx, url)
+	if err != nil {
+		s.Error = err
+		return 
```

Comment:
I am pretty sure this would return nil instead of err. I believe you've been hit by a special case: the second returned value variable created at line 171 shed the err variable of the named return. 

I recommend not using a named return in this case.

Reply:
Awesome! Fixing it!